tfmigrate/test_xmv_action_test.go

```diff
@@ -0,0 +1,165 @@
+
+	actions := []StateAction{
+		NewStateXMvAction("aws_security_group.*", "module.sg_module[\"$1\"].aws_security_group.sg"),
+	}
+
+	m := NewStateMigrator(tf.Dir(), "default", actions, &MigratorOption{}, false)
+	err = m.Plan(ctx)
+	if err != nil {
+		t.Fatalf("failed to run migrator plan: %s", err)
+	}
+}
+
+func TestGetNrOfWildcard(t *testing.T) {
+	cases := []struct {
+		desc        string
+		resource    *StateXMvAction
+		nrWildcards int
+	}{
+		{
+			desc:        "Simple resource no wildcardChar",
+			resource:    NewStateXMvAction("aws_security_group.foo", "aws_security_group.foo2"),
+			nrWildcards: 0,
+		},
+		{
+			desc:        "Simple wildcardChar for a resource",
+			resource:    NewStateXMvAction("aws_security_group.*", "aws_security_group.$1"),
+			nrWildcards: 1,
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.desc, func(t *testing.T) {
+			got := tc.resource.nrOfWildcards()
+			if got != tc.nrWildcards {
+				t.Errorf("Number of wildcards for %d is not expected %s", got, tc.resource)
+			}
+		})
+	}
+}
+
+func TestSingleResourceStateMoves(t *testing.T) {
+	cases := []struct {
+		desc            string
+		stateList       []string
+		inputXMvAction  *StateXMvAction
+		outputMvActions []*StateMvAction
+	}{
+		{
+			desc:      "Simple resource no wildcardChar",
+			stateList: nil,
+			inputXMvAction: &StateXMvAction{
+				source:      "aws_security_group.foo",
+				destination: "aws_security_group.foo2",
+			},
+			outputMvActions: []*StateMvAction{
+				{
+					source:      "aws_security_group.foo",
+					destination: "aws_security_group.foo2",
+				},
+			},
+		},
+		{
+			desc:      "Simple resource with wildcardChar",
+			stateList: []string{"aws_security_group.foo"},
+			inputXMvAction: &StateXMvAction{
+				source:      "aws_security_group.*",
+				destination: "module.security_group[\"$1\"].sg",
+			},
+			outputMvActions: []*StateMvAction{
+				{
+					source:      "aws_security_group.foo",
+					destination: "module.security_group[\"foo\"].sg",
+				},
+			},
+		},
+		{
+			desc:      "Simple module name refactor with wildcardChar",
+			stateList: []string{"module.security_group[\"foo\"].sg"},
+			inputXMvAction: &StateXMvAction{
+				source:      "module.security_group[\"*\"].sg",
+				destination: "module.sg[\"$1\"].sg",
+			},
+			outputMvActions: []*StateMvAction{
+				{
+					source:      "module.security_group[\"foo\"].sg",
+					destination: "module.sg[\"foo\"].sg",
+				},
+			},
+		},
+		{
+			desc:      "No matching resources in state",
+			stateList: []string{"aws_vpc.foo"},
+			inputXMvAction: &StateXMvAction{
+				source:      "aws_security_group.*",
+				destination: "module.security_group[\"$1\"].sg",
+			},
+			outputMvActions: []*StateMvAction{},
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.desc, func(t *testing.T) {
+			got, err := tc.inputXMvAction.getStateMvActionsForStateList(tc.stateList)
+			// Should be location agnostic
+			if err != nil {
+				t.Errorf("Encountered error %v", err)
+			}
```

Comment:
What errors may be returned? Is it possible to test for error cases?

Reply:
I think this is the trickiest question. 

W.r.t what errors may be returned it is only the one that originates from `makeSrcRegex`:

If there is an error in the regexp the compilation fails (e.g. unbalanced round brackets). I don't think it should be possible to hit this since we only introduces braces like that in a balanced way and in the original expression Meta characters are quoted. 

One way to simulate the error is by changing `const matchWildcardRegex = "(.*)"` to `const matchWildcardRegex = "(.*"` and running the unittests. For now I just documented it. If you have suggestions on how I can do that in a test that is welcome but I want to avoid passing `matchWildcardRegex` for the purpose of this test. 

Perhaps this is an indication that restructuring is required?  For example if instead a map is used that maps a char to a regex passing that map into getStateMvActionsForStateList might be acceptable. Generally I don't like to over-design early on because if other type of matching is needed it could be that the map approach still wouldn't work.

I'm open for any suggestion.

Reply:
I agree that changing the `matchWildcardRegex` only for testing makes no sense.

If an input regular expression is a fixed string, it is possible to use `regexp.MustCompile`, which will panic without returning an error.
https://pkg.go.dev/regexp#MustCompile

However, in this case, I don't think it is appropriate to use panic at runtime because it receives user input.

My proposal is to use `t.Fatalf` instead of `t.Errorf`; it will fail the test case immediately when something is wrong and stop further assertions. 
https://pkg.go.dev/testing#T.Fatalf